### Case Study - Uber Data Analysis

The data of a driver's uber trips are available for year 2016.Give some useful insights about the trip behaviour of a Uber driver.



### Dataset -
The dataset contains Start Date, End Date, Start Location, End Location, Miles Driven and Purpose of drive (Business, Personal, Meals, Errands, Meetings, Customer Support etc.)



#### 1. Data profiling
#### 2. group by function
#### 3. Apply function
#### 4. DateTime operations

In [ ]:
#Import the libraries

import pandas as pd
import numpy as np
import datetime as dt

In [ ]:
# Read the data
url='https://drive.google.com/file/d/17M30K1_cnKpFbfWc_CGY9c_58RfoLgUX/view?usp=share_link'
url='https://drive.google.com/uc?id=' + url.split('/')[-2]
df=pd.read_csv(url)
#df = pd.read_csv('/content/My Uber Drives - 2016.csv')


In [ ]:
# View first n rows of the data; n=7

df.head(7)

START_DATE*       END_DATE* CATEGORY*           START*            STOP*  \
0  1/1/2016 21:11  1/1/2016 21:17  Business      Fort Pierce      Fort Pierce   
1   1/2/2016 1:25   1/2/2016 1:37  Business      Fort Pierce      Fort Pierce   
2  1/2/2016 20:25  1/2/2016 20:38  Business      Fort Pierce      Fort Pierce   
3  1/5/2016 17:31  1/5/2016 17:45  Business      Fort Pierce      Fort Pierce   
4  1/6/2016 14:42  1/6/2016 15:49  Business      Fort Pierce  West Palm Beach   
5  1/6/2016 17:15  1/6/2016 17:19  Business  West Palm Beach  West Palm Beach   
6  1/6/2016 17:30  1/6/2016 17:35  Business  West Palm Beach       Palm Beach   

   MILES*         PURPOSE*  
0     5.1   Meal/Entertain  
1     5.0              NaN  
2     4.8  Errand/Supplies  
3     4.7          Meeting  
4    63.7   Customer Visit  
5     4.3   Meal/Entertain  
6     7.1          Meeting

In [ ]:
# View the last 5 rows of data
df.tail(5)

START_DATE*         END_DATE* CATEGORY*            START*  \
1151  12/31/2016 13:24  12/31/2016 13:42  Business           Kar?chi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   
1153  12/31/2016 21:32  12/31/2016 21:50  Business        Katunayake   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   
1155            Totals               NaN       NaN               NaN   

                 STOP*   MILES*        PURPOSE*  
1151  Unknown Location      3.9  Temporary Site  
1152  Unknown Location     16.2         Meeting  
1153           Gampaha      6.4  Temporary Site  
1154         Ilukwatta     48.2  Temporary Site  
1155               NaN  12204.7             NaN

In [ ]:
# Understand shape of the data

df.shape

(1156, 7)

##### There are 1156 rows and 7 columns in the dataset.

In [ ]:
# Size of Data Frame
df.size

8092

In [ ]:
# Number of Null Records
df.isnull().sum()

START_DATE*      0
END_DATE*        1
CATEGORY*        1
START*           1
STOP*            1
MILES*           0
PURPOSE*       503
dtype: int64

In [ ]:
#Null Values
df_null=df[df.isnull()]
df_null

START_DATE* END_DATE* CATEGORY* START* STOP*  MILES* PURPOSE*
0            NaN       NaN       NaN    NaN   NaN     NaN      NaN
1            NaN       NaN       NaN    NaN   NaN     NaN      NaN
2            NaN       NaN       NaN    NaN   NaN     NaN      NaN
3            NaN       NaN       NaN    NaN   NaN     NaN      NaN
4            NaN       NaN       NaN    NaN   NaN     NaN      NaN
...          ...       ...       ...    ...   ...     ...      ...
1151         NaN       NaN       NaN    NaN   NaN     NaN      NaN
1152         NaN       NaN       NaN    NaN   NaN     NaN      NaN
1153         NaN       NaN       NaN    NaN   NaN     NaN      NaN
1154         NaN       NaN       NaN    NaN   NaN     NaN      NaN
1155         NaN       NaN       NaN    NaN   NaN     NaN      NaN

[1156 rows x 7 columns]

In [ ]:
# Not Null Values
df_notnull = df[~df.isnull()]
df_notnull

START_DATE*         END_DATE* CATEGORY*            START*  \
0       1/1/2016 21:11    1/1/2016 21:17  Business       Fort Pierce   
1        1/2/2016 1:25     1/2/2016 1:37  Business       Fort Pierce   
2       1/2/2016 20:25    1/2/2016 20:38  Business       Fort Pierce   
3       1/5/2016 17:31    1/5/2016 17:45  Business       Fort Pierce   
4       1/6/2016 14:42    1/6/2016 15:49  Business       Fort Pierce   
...                ...               ...       ...               ...   
1151  12/31/2016 13:24  12/31/2016 13:42  Business           Kar?chi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   
1153  12/31/2016 21:32  12/31/2016 21:50  Business        Katunayake   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   
1155            Totals               NaN       NaN               NaN   

                 STOP*   MILES*         PURPOSE*  
0          Fort Pierce      5.1   Meal/Entertain  
1          Fort Pierce      5.0              NaN  
2          Fort Pierce      4.8  Errand/Supplies  
3          Fort Pierce      4.7          Meeting  
4      West Palm Beach     63.7   Customer Visit  
...                ...      ...              ...  
1151  Unknown Location      3.9   Temporary Site  
1152  Unknown Location     16.2          Meeting  
1153           Gampaha      6.4   Temporary Site  
1154         Ilukwatta     48.2   Temporary Site  
1155               NaN  12204.7              NaN  

[1156 rows x 7 columns]

In [ ]:
# Show Records with missing values for column = Purpose

df[df['PURPOSE*'].isnull()]

START_DATE*         END_DATE* CATEGORY*            START*  \
1        1/2/2016 1:25     1/2/2016 1:37  Business       Fort Pierce   
32      1/19/2016 9:09    1/19/2016 9:23  Business       Whitebridge   
85      2/9/2016 10:54    2/9/2016 11:07  Personal       Whitebridge   
86      2/9/2016 11:43    2/9/2016 11:50  Personal        Northwoods   
87      2/9/2016 13:36    2/9/2016 13:52  Personal        Tanglewood   
...                ...               ...       ...               ...   
1066  12/19/2016 14:37  12/19/2016 14:50  Business  Unknown Location   
1069  12/19/2016 19:05  12/19/2016 19:17  Business         Islamabad   
1071   12/20/2016 8:49   12/20/2016 9:24  Business  Unknown Location   
1143  12/29/2016 20:53  12/29/2016 21:42  Business           Kar?chi   
1155            Totals               NaN       NaN               NaN   

                    STOP*   MILES* PURPOSE*  
1             Fort Pierce      5.0      NaN  
32    Lake Wellingborough      7.2      NaN  
85             Northwoods      5.3      NaN  
86             Tanglewood      3.0      NaN  
87                Preston      5.1      NaN  
...                   ...      ...      ...  
1066     Unknown Location      5.4      NaN  
1069     Unknown Location      2.2      NaN  
1071           Rawalpindi     12.0      NaN  
1143     Unknown Location      6.4      NaN  
1155                  NaN  12204.7      NaN  

[503 rows x 7 columns]

In [ ]:
# Records with non null values in column = 'Purpose'

df[~df['PURPOSE*'].isnull()]

START_DATE*         END_DATE* CATEGORY*            START*  \
0       1/1/2016 21:11    1/1/2016 21:17  Business       Fort Pierce   
2       1/2/2016 20:25    1/2/2016 20:38  Business       Fort Pierce   
3       1/5/2016 17:31    1/5/2016 17:45  Business       Fort Pierce   
4       1/6/2016 14:42    1/6/2016 15:49  Business       Fort Pierce   
5       1/6/2016 17:15    1/6/2016 17:19  Business   West Palm Beach   
...                ...               ...       ...               ...   
1150   12/31/2016 1:07   12/31/2016 1:14  Business           Kar?chi   
1151  12/31/2016 13:24  12/31/2016 13:42  Business           Kar?chi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   
1153  12/31/2016 21:32  12/31/2016 21:50  Business        Katunayake   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   

                 STOP*  MILES*         PURPOSE*  
0          Fort Pierce     5.1   Meal/Entertain  
2          Fort Pierce     4.8  Errand/Supplies  
3          Fort Pierce     4.7          Meeting  
4      West Palm Beach    63.7   Customer Visit  
5      West Palm Beach     4.3   Meal/Entertain  
...                ...     ...              ...  
1150           Kar?chi     0.7          Meeting  
1151  Unknown Location     3.9   Temporary Site  
1152  Unknown Location    16.2          Meeting  
1153           Gampaha     6.4   Temporary Site  
1154         Ilukwatta    48.2   Temporary Site  

[653 rows x 7 columns]

In [ ]:
# Number of NonNull records in Purpose column
(~df['PURPOSE*'].isnull()).sum()

653

### Renaming Columns

In [ ]:
df.rename( columns = {'START_DATE*':'START_DATE', 'END_DATE*':'END_DATE', 'CATEGORY*': 'CATEGORY', 'START*':'START', 'STOP*':'STOP', 'MILES*':'MILES', 'PURPOSE*':'PURPOSE'}, inplace=True)
df.head()

START_DATE        END_DATE  CATEGORY        START             STOP  \
0  1/1/2016 21:11  1/1/2016 21:17  Business  Fort Pierce      Fort Pierce   
1   1/2/2016 1:25   1/2/2016 1:37  Business  Fort Pierce      Fort Pierce   
2  1/2/2016 20:25  1/2/2016 20:38  Business  Fort Pierce      Fort Pierce   
3  1/5/2016 17:31  1/5/2016 17:45  Business  Fort Pierce      Fort Pierce   
4  1/6/2016 14:42  1/6/2016 15:49  Business  Fort Pierce  West Palm Beach   

   MILES          PURPOSE  
0    5.1   Meal/Entertain  
1    5.0              NaN  
2    4.8  Errand/Supplies  
3    4.7          Meeting  
4   63.7   Customer Visit

#### Filtering Dataframes

In [ ]:
# Entries where purpose is null

df[df.PURPOSE.isnull()]

START_DATE          END_DATE  CATEGORY             START  \
1        1/2/2016 1:25     1/2/2016 1:37  Business       Fort Pierce   
32      1/19/2016 9:09    1/19/2016 9:23  Business       Whitebridge   
85      2/9/2016 10:54    2/9/2016 11:07  Personal       Whitebridge   
86      2/9/2016 11:43    2/9/2016 11:50  Personal        Northwoods   
87      2/9/2016 13:36    2/9/2016 13:52  Personal        Tanglewood   
...                ...               ...       ...               ...   
1066  12/19/2016 14:37  12/19/2016 14:50  Business  Unknown Location   
1069  12/19/2016 19:05  12/19/2016 19:17  Business         Islamabad   
1071   12/20/2016 8:49   12/20/2016 9:24  Business  Unknown Location   
1143  12/29/2016 20:53  12/29/2016 21:42  Business           Kar?chi   
1155            Totals               NaN       NaN               NaN   

                     STOP    MILES PURPOSE  
1             Fort Pierce      5.0     NaN  
32    Lake Wellingborough      7.2     NaN  
85             Northwoods      5.3     NaN  
86             Tanglewood      3.0     NaN  
87                Preston      5.1     NaN  
...                   ...      ...     ...  
1066     Unknown Location      5.4     NaN  
1069     Unknown Location      2.2     NaN  
1071           Rawalpindi     12.0     NaN  
1143     Unknown Location      6.4     NaN  
1155                  NaN  12204.7     NaN  

[503 rows x 7 columns]

In [ ]:
#Inverting the selection
df[df.PURPOSE.notnull()]

START_DATE          END_DATE  CATEGORY             START  \
0       1/1/2016 21:11    1/1/2016 21:17  Business       Fort Pierce   
2       1/2/2016 20:25    1/2/2016 20:38  Business       Fort Pierce   
3       1/5/2016 17:31    1/5/2016 17:45  Business       Fort Pierce   
4       1/6/2016 14:42    1/6/2016 15:49  Business       Fort Pierce   
5       1/6/2016 17:15    1/6/2016 17:19  Business   West Palm Beach   
...                ...               ...       ...               ...   
1150   12/31/2016 1:07   12/31/2016 1:14  Business           Kar?chi   
1151  12/31/2016 13:24  12/31/2016 13:42  Business           Kar?chi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   
1153  12/31/2016 21:32  12/31/2016 21:50  Business        Katunayake   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   

                  STOP  MILES          PURPOSE  
0          Fort Pierce    5.1   Meal/Entertain  
2          Fort Pierce    4.8  Errand/Supplies  
3          Fort Pierce    4.7          Meeting  
4      West Palm Beach   63.7   Customer Visit  
5      West Palm Beach    4.3   Meal/Entertain  
...                ...    ...              ...  
1150           Kar?chi    0.7          Meeting  
1151  Unknown Location    3.9   Temporary Site  
1152  Unknown Location   16.2          Meeting  
1153           Gampaha    6.4   Temporary Site  
1154         Ilukwatta   48.2   Temporary Site  

[653 rows x 7 columns]

####Filtering Dataframes - 2

In [ ]:
# print df where StOP is Fort Pierce

df[df.STOP=='Fort Pierce']

START_DATE        END_DATE  CATEGORY        START         STOP  MILES  \
0  1/1/2016 21:11  1/1/2016 21:17  Business  Fort Pierce  Fort Pierce    5.1   
1   1/2/2016 1:25   1/2/2016 1:37  Business  Fort Pierce  Fort Pierce    5.0   
2  1/2/2016 20:25  1/2/2016 20:38  Business  Fort Pierce  Fort Pierce    4.8   
3  1/5/2016 17:31  1/5/2016 17:45  Business  Fort Pierce  Fort Pierce    4.7   

           PURPOSE  
0   Meal/Entertain  
1              NaN  
2  Errand/Supplies  
3          Meeting

In [ ]:
df[df.STOP!='Fort Pierce']

START_DATE          END_DATE  CATEGORY             START  \
4       1/6/2016 14:42    1/6/2016 15:49  Business       Fort Pierce   
5       1/6/2016 17:15    1/6/2016 17:19  Business   West Palm Beach   
6       1/6/2016 17:30    1/6/2016 17:35  Business   West Palm Beach   
7       1/7/2016 13:27    1/7/2016 13:33  Business              Cary   
8       1/10/2016 8:05    1/10/2016 8:25  Business              Cary   
...                ...               ...       ...               ...   
1151  12/31/2016 13:24  12/31/2016 13:42  Business           Kar?chi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   
1153  12/31/2016 21:32  12/31/2016 21:50  Business        Katunayake   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   
1155            Totals               NaN       NaN               NaN   

                  STOP    MILES         PURPOSE  
4      West Palm Beach     63.7  Customer Visit  
5      West Palm Beach      4.3  Meal/Entertain  
6           Palm Beach      7.1         Meeting  
7                 Cary      0.8         Meeting  
8          Morrisville      8.3         Meeting  
...                ...      ...             ...  
1151  Unknown Location      3.9  Temporary Site  
1152  Unknown Location     16.2         Meeting  
1153           Gampaha      6.4  Temporary Site  
1154         Ilukwatta     48.2  Temporary Site  
1155               NaN  12204.7             NaN  

[1152 rows x 7 columns]

####Explore the details from the MILES column


In [ ]:
# Show top 10 rides

df_Sortedbymiles = df.sort_values(by = 'MILES', ascending = False).head(10)
df_Sortedbymiles

START_DATE          END_DATE  CATEGORY             START  \
1155            Totals               NaN       NaN               NaN   
269    3/25/2016 16:52   3/25/2016 22:22  Business             Latta   
270    3/25/2016 22:54    3/26/2016 1:39  Business      Jacksonville   
881   10/30/2016 15:22  10/30/2016 18:23  Business         Asheville   
776    9/27/2016 21:01    9/28/2016 2:37  Business  Unknown Location   
546    7/14/2016 16:39   7/14/2016 20:05  Business       Morrisville   
559    7/17/2016 12:20   7/17/2016 15:25  Personal             Boone   
297     4/2/2016 19:38    4/2/2016 22:36  Business      Jacksonville   
299      4/3/2016 2:00     4/3/2016 4:16  Business          Florence   
727    8/27/2016 16:15   8/27/2016 19:13  Business  Unknown Location   

                  STOP    MILES         PURPOSE  
1155               NaN  12204.7             NaN  
269       Jacksonville    310.3  Customer Visit  
270          Kissimmee    201.0         Meeting  
881             Mebane    195.9             NaN  
776   Unknown Location    195.6             NaN  
546         Banner Elk    195.3             NaN  
559               Cary    180.2         Commute  
297          Ridgeland    174.2  Customer Visit  
299               Cary    159.3         Meeting  
727   Unknown Location    156.9             NaN

In [ ]:
# Show the row that has max miles

df_Sortedbymiles.head(1)

START_DATE END_DATE CATEGORY START STOP    MILES PURPOSE
1155     Totals      NaN      NaN   NaN  NaN  12204.7     NaN

In [ ]:
#summary
df.describe()

MILES
count   1156.000000
mean      21.115398
std      359.299007
min        0.500000
25%        2.900000
50%        6.000000
75%       10.400000
max    12204.700000

In [ ]:
#summary

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1156 entries, 0 to 1155
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   START_DATE  1156 non-null   object 
 1   END_DATE    1155 non-null   object 
 2   CATEGORY    1155 non-null   object 
 3   START       1155 non-null   object 
 4   STOP        1155 non-null   object 
 5   MILES       1156 non-null   float64
 6   PURPOSE     653 non-null    object 
dtypes: float64(1), object(6)
memory usage: 63.3+ KB


####Removal of Null Values

In [ ]:
#creating duplicate set

df_1 = df.copy()

In [ ]:
#dropping null values
df_1.dropna(axis=0, how = 'any', inplace=True)

In [ ]:
#get the shape of revised dataframe

df_1.shape

(653, 7)

### PANDAS PROFILING


In [ ]:
! pip install https://github.com/pandas-profiling/pandas-profiling/archive/master.zip

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     / 22.0 MB 17.6 MB/s 0:00:02
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 102.7/102.7 KB 9.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 679.5/679.5 KB 43.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.9/9.9 MB 83.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.7/4.7 MB 18.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 296.5/296.5 KB 31.9 MB/s eta 0:00:00
  Created wheel for pandas-profiling: filename=pandas_profiling-dev-py2.py3-none-any.whl size=328711 sha256=fa458daabf61410c183b64043d31e897bd5fc4f67162ea521b76f5ee4abebf7c
  Stored in directory: /tmp/pip-ephem-wheel-cache-33k8h57q/wheels/64/b6/85/dfc808b23666a5910371784e349d28818006ff63ed9cfeca59
  Created wheel for htmlmin: filename=htmlmin-0.1.12-py3-none-any.whl size=27098 sha25

In [ ]:
import pandas_profiling as pp


In [ ]:
profile=pp.ProfileReport(df_1, title="Pandas Profiling Report")

In [ ]:
profile.to_file('output.html')

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

###Lets explore the data parameter wise -

##1. Destination



In [ ]:
df[['START', 'STOP']]

START              STOP
0          Fort Pierce       Fort Pierce
1          Fort Pierce       Fort Pierce
2          Fort Pierce       Fort Pierce
3          Fort Pierce       Fort Pierce
4          Fort Pierce   West Palm Beach
...                ...               ...
1151           Kar?chi  Unknown Location
1152  Unknown Location  Unknown Location
1153        Katunayake           Gampaha
1154           Gampaha         Ilukwatta
1155               NaN               NaN

[1156 rows x 2 columns]

##2.Time - (hour of the day, day of week, month of year)


In [ ]:
df_1=df_1[df_1['START_DATE']!='Totals']
df_1=df_1[df_1['END_DATE']!='Totals']
df_1.head(2)

START_DATE        END_DATE  CATEGORY        START         STOP  MILES  \
0  1/1/2016 21:11  1/1/2016 21:17  Business  Fort Pierce  Fort Pierce    5.1   
2  1/2/2016 20:25  1/2/2016 20:38  Business  Fort Pierce  Fort Pierce    4.8   

           PURPOSE  
0   Meal/Entertain  
2  Errand/Supplies

##3. Categories

In [ ]:
df['CATEGORY']

0       Business
1       Business
2       Business
3       Business
4       Business
          ...   
1151    Business
1152    Business
1153    Business
1154    Business
1155         NaN
Name: CATEGORY, Length: 1156, dtype: object

In [ ]:
df['CATEGORY'].unique()

array(['Business', 'Personal', nan], dtype=object)

##4.Purpose

In [ ]:
df['PURPOSE'].unique()

array(['Meal/Entertain', nan, 'Errand/Supplies', 'Meeting',
       'Customer Visit', 'Temporary Site', 'Between Offices',
       'Charity ($)', 'Commute', 'Moving', 'Airport/Travel'], dtype=object)

### Grouping two parameters to get more insights


###1. Understanding  the start and stop points

# names of unique start points

In [ ]:
df['START'].unique()

array(['Fort Pierce', 'West Palm Beach', 'Cary', 'Jamaica', 'New York',
       'Elmhurst', 'Midtown', 'East Harlem', 'Flatiron District',
       'Midtown East', 'Hudson Square', 'Lower Manhattan',
       "Hell's Kitchen", 'Downtown', 'Gulfton', 'Houston', 'Eagan Park',
       'Morrisville', 'Durham', 'Farmington Woods', 'Whitebridge',
       'Lake Wellingborough', 'Fayetteville Street', 'Raleigh',
       'Hazelwood', 'Fairmont', 'Meredith Townes', 'Apex', 'Chapel Hill',
       'Northwoods', 'Edgehill Farms', 'Tanglewood', 'Preston',
       'Eastgate', 'East Elmhurst', 'Jackson Heights', 'Long Island City',
       'Katunayaka', 'Unknown Location', 'Colombo', 'Nugegoda',
       'Islamabad', 'R?walpindi', 'Noorpur Shahan', 'Heritage Pines',
       'Westpark Place', 'Waverly Place', 'Wayne Ridge', 'Weston',
       'East Austin', 'West University', 'South Congress', 'The Drag',
       'Congress Ave District', 'Red River District', 'Georgian Acres',
       'North Austin', 'Coxville', 'Conven

In [ ]:
#count if unique start points

df['START'].nunique()

177

In [ ]:
# unique destinations

df['STOP'].unique()

array(['Fort Pierce', 'West Palm Beach', 'Palm Beach', 'Cary',
       'Morrisville', 'New York', 'Queens', 'East Harlem', 'NoMad',
       'Midtown', 'Midtown East', 'Hudson Square', 'Lower Manhattan',
       "Hell's Kitchen", 'Queens County', 'Gulfton', 'Downtown',
       'Houston', 'Jamestown Court', 'Durham', 'Whitebridge',
       'Lake Wellingborough', 'Raleigh', 'Umstead', 'Hazelwood',
       'Westpark Place', 'Meredith Townes', 'Leesville Hollow', 'Apex',
       'Chapel Hill', 'Williamsburg Manor', 'Macgregor Downs',
       'Edgehill Farms', 'Northwoods', 'Tanglewood', 'Preston',
       'Walnut Terrace', 'Jackson Heights', 'East Elmhurst',
       'Midtown West', 'Long Island City', 'Jamaica', 'Unknown Location',
       'Colombo', 'Nugegoda', 'Katunayaka', 'Islamabad', 'R?walpindi',
       'Noorpur Shahan', 'Heritage Pines', 'Waverly Place', 'Wayne Ridge',
       'Depot Historic District', 'Weston', 'West University',
       'South Congress', 'Arts District', 'Congress Ave District

In [ ]:
#count of unique stop points
df['STOP'].nunique()

188

In [ ]:
#stations that appear in both Start and Stop
df[df['START'].isin(df['STOP'])]['START']

0            Fort Pierce
1            Fort Pierce
2            Fort Pierce
3            Fort Pierce
4            Fort Pierce
              ...       
1150             Kar?chi
1151             Kar?chi
1152    Unknown Location
1154             Gampaha
1155                 NaN
Name: START, Length: 1131, dtype: object

In [ ]:
#top 10 start points by popularity
df_2=df.groupby(by='START').count()['START_DATE']
df_2.sort_values(axis=0, ascending=False).head(10)

START
Cary                201
Unknown Location    148
Morrisville          85
Whitebridge          68
Islamabad            57
Durham               37
Lahore               36
Raleigh              28
Kar?chi              27
Apex                 17
Name: START_DATE, dtype: int64

In [ ]:
#top 10 stop points by popularity

df_3=df.groupby(by='STOP').count()['START_DATE']
df_3.sort_values(axis=0, ascending=False).head(10)

STOP
Cary                203
Unknown Location    149
Morrisville          84
Whitebridge          65
Islamabad            58
Durham               36
Lahore               36
Raleigh              29
Kar?chi              26
Apex                 17
Name: START_DATE, dtype: int64

In [ ]:
#cases where start and stop locations are same

df[df['START']==df['STOP']]

START_DATE          END_DATE  CATEGORY             START  \
0       1/1/2016 21:11    1/1/2016 21:17  Business       Fort Pierce   
1        1/2/2016 1:25     1/2/2016 1:37  Business       Fort Pierce   
2       1/2/2016 20:25    1/2/2016 20:38  Business       Fort Pierce   
3       1/5/2016 17:31    1/5/2016 17:45  Business       Fort Pierce   
5       1/6/2016 17:15    1/6/2016 17:19  Business   West Palm Beach   
...                ...               ...       ...               ...   
1147  12/30/2016 15:41  12/30/2016 16:03  Business           Kar?chi   
1148  12/30/2016 16:45  12/30/2016 17:08  Business           Kar?chi   
1149  12/30/2016 23:06  12/30/2016 23:10  Business           Kar?chi   
1150   12/31/2016 1:07   12/31/2016 1:14  Business           Kar?chi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   

                  STOP  MILES          PURPOSE  
0          Fort Pierce    5.1   Meal/Entertain  
1          Fort Pierce    5.0              NaN  
2          Fort Pierce    4.8  Errand/Supplies  
3          Fort Pierce    4.7          Meeting  
5      West Palm Beach    4.3   Meal/Entertain  
...                ...    ...              ...  
1147           Kar?chi    4.6  Errand/Supplies  
1148           Kar?chi    4.6          Meeting  
1149           Kar?chi    0.8   Customer Visit  
1150           Kar?chi    0.7          Meeting  
1152  Unknown Location   16.2          Meeting  

[288 rows x 7 columns]

In [ ]:
#Favorite starting point wrt the total miles covered

df.groupby(by='START').MILES.sum().sort_values(ascending=False)

START
Unknown Location         1976.5
Cary                     1791.3
Morrisville               671.7
Raleigh                   433.0
Islamabad                 401.2
                          ...  
South Berkeley              0.9
Congress Ave District       0.8
Sunnyside                   0.7
Medical Centre              0.7
Soho                        0.5
Name: MILES, Length: 177, dtype: float64

In [ ]:
#START POINT BY MAXIMUM MILES COVERED
df_1[df_1['MILES']==df_1.MILES.max()][['START', 'MILES']]

START  MILES
269  Latta  310.3

In [ ]:
#dropping Unknown locations
df_1 = df_1[df['START']!='Unknown Location']
df_1 = df_1[df['STOP']!='Unknown Location']


<ipython-input-95-1d7b693bb6de>:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_1 = df_1[df['START']!='Unknown Location']
<ipython-input-95-1d7b693bb6de>:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_1 = df_1[df['START']!='Unknown Location']


In [ ]:
df_1.pivot_table(values='MILES', index=['START', 'STOP'], aggfunc='sum').sort_values(by='MILES', ascending=False).head(1)

MILES
START STOP         
Cary  Durham  312.3

In [ ]:
df_1.pivot_table(values='MILES', index=['START', 'STOP'], aggfunc='count').sort_values(by='MILES', ascending=False).head(1)

MILES
START STOP              
Cary  Morrisville     52

## 2. Manipulating date & time objects

In [ ]:
df.head()

START_DATE        END_DATE  CATEGORY        START             STOP  \
0  1/1/2016 21:11  1/1/2016 21:17  Business  Fort Pierce      Fort Pierce   
1   1/2/2016 1:25   1/2/2016 1:37  Business  Fort Pierce      Fort Pierce   
2  1/2/2016 20:25  1/2/2016 20:38  Business  Fort Pierce      Fort Pierce   
3  1/5/2016 17:31  1/5/2016 17:45  Business  Fort Pierce      Fort Pierce   
4  1/6/2016 14:42  1/6/2016 15:49  Business  Fort Pierce  West Palm Beach   

   MILES          PURPOSE  
0    5.1   Meal/Entertain  
1    5.0              NaN  
2    4.8  Errand/Supplies  
3    4.7          Meeting  
4   63.7   Customer Visit

In [ ]:
df_1=df_1[df_1['START_DATE']!='Totals']
df_1=df_1[df_1['END_DATE']!='Totals']

In [ ]:
# Create columns by converting the start and end date into a datatime format
df_1['START_DATE'] = pd.to_datetime(df_1['START_DATE'])
df_1['END_DATE'] = pd.to_datetime(df_1['END_DATE'])
df_1.head()



START_DATE            END_DATE  CATEGORY            START  \
0 2016-01-01 21:11:00 2016-01-01 21:17:00  Business      Fort Pierce   
2 2016-01-02 20:25:00 2016-01-02 20:38:00  Business      Fort Pierce   
3 2016-01-05 17:31:00 2016-01-05 17:45:00  Business      Fort Pierce   
4 2016-01-06 14:42:00 2016-01-06 15:49:00  Business      Fort Pierce   
5 2016-01-06 17:15:00 2016-01-06 17:19:00  Business  West Palm Beach   

              STOP  MILES          PURPOSE  
0      Fort Pierce    5.1   Meal/Entertain  
2      Fort Pierce    4.8  Errand/Supplies  
3      Fort Pierce    4.7          Meeting  
4  West Palm Beach   63.7   Customer Visit  
5  West Palm Beach    4.3   Meal/Entertain

In [ ]:
# Which month did he get most drives  ?
df_1['Month'] = pd.DatetimeIndex(df_1['START_DATE']).month
df_1.head()

START_DATE            END_DATE  CATEGORY            START  \
0 2016-01-01 21:11:00 2016-01-01 21:17:00  Business      Fort Pierce   
2 2016-01-02 20:25:00 2016-01-02 20:38:00  Business      Fort Pierce   
3 2016-01-05 17:31:00 2016-01-05 17:45:00  Business      Fort Pierce   
4 2016-01-06 14:42:00 2016-01-06 15:49:00  Business      Fort Pierce   
5 2016-01-06 17:15:00 2016-01-06 17:19:00  Business  West Palm Beach   

              STOP  MILES          PURPOSE  Month  
0      Fort Pierce    5.1   Meal/Entertain      1  
2      Fort Pierce    4.8  Errand/Supplies      1  
3      Fort Pierce    4.7          Meeting      1  
4  West Palm Beach   63.7   Customer Visit      1  
5  West Palm Beach    4.3   Meal/Entertain      1

In [ ]:
df_1['Month'].value_counts()

12    101
6      73
3      71
2      68
1      59
11     59
4      50
5      46
7      46
10     17
8       8
Name: Month, dtype: int64

In [ ]:
# Getting the average distance covered each month

df_1.groupby(by='Month').mean()

MILES
Month           
1       8.486441
2       8.186765
3      20.505634
4      21.898000
5       7.793478
6       7.410959
7      10.615217
8       4.912500
10     26.764706
11     10.710169
12      6.097030

In [ ]:
# Which day did he get most drives  ?
df_1['Day'] = pd.DatetimeIndex(df_1['START_DATE']).day
df_1['Day'].value_counts()

4     28
12    27
29    27
1     25
8     25
28    25
19    24
18    24
7     23
6     22
5     21
3     20
26    20
14    20
11    20
10    19
27    19
25    19
20    18
2     18
21    18
13    17
22    17
24    17
15    15
16    15
30    12
17    12
9     11
23    11
31     9
Name: Day, dtype: int64

In [ ]:
# Question1: How many miles was earned per category and purpose ?

df_1.groupby(by=['CATEGORY', 'PURPOSE']).MILES.sum()

CATEGORY  PURPOSE        
Business  Airport/Travel        4.1
          Between Offices     197.0
          Customer Visit     2064.7
          Errand/Supplies     441.9
          Meal/Entertain      878.5
          Meeting            2581.7
          Temporary Site      356.6
Personal  Charity ($)          15.1
          Commute             180.2
          Moving               18.2
Name: MILES, dtype: float64